src/H5FDsubfiling/H5FDsubfiling.c

```diff
@@ -2792,212 +3027,268 @@ init_indep_io(subfiling_context_t *sf_context, int64_t file_offset, size_t io_ne
-            goto done;
-        }
+        iovec_depth      = cur_max_iovec_depth;
+        num_full_stripes = iovec_depth;
 
-        if (total_bytes < data_size) {
-            int64_t num_full_stripes = iovec_depth;
+        if (subfile_idx == first_subfile_idx) {
+            is_first = true;
 
-            if (k == first_subfile) {
-                is_first = true;
+            /*
+             * Add partial segment length if not
+             * starting on a stripe boundary
+             */
+            if (start_length < stripe_size) {
+                subfile_bytes += start_length;
+                num_full_stripes--;
+            }
+        }
 
-                /*
-                 * Add partial segment length if not
-                 * starting on a stripe boundary
-                 */
-                if (start_length < stripe_size) {
-                    subfile_bytes += start_length;
+        if (subfile_idx == last_subfile_idx) {
+            is_last = true;
+
+            /*
+             * Add partial segment length if not
+             * ending on a stripe boundary
+             */
+            if (final_length < stripe_size) {
+                subfile_bytes += final_length;
+                if (num_full_stripes)
                     num_full_stripes--;
-                }
             }
+        }
 
-            if (k == last_subfile) {
-                is_last = true;
+        /* Account for subfiles with uniform segments */
+        if (!is_first && !is_last) {
+            bool thin_uniform_section = false;
 
+            if (last_subfile_idx >= first_subfile_idx) {
                 /*
-                 * Add partial segment length if not
-                 * ending on a stripe boundary
+                 * When a subfile has an index value that is greater
+                 * than both the starting subfile and ending subfile
+                 * indices, it is a "thinner" section with a smaller
+                 * I/O vector depth.
                  */
-                if (final_length < stripe_size) {
-                    subfile_bytes += final_length;
-                    if (num_full_stripes)
-                        num_full_stripes--;
-                }
+                thin_uniform_section = (subfile_idx > first_subfile_idx) && (subfile_idx > last_subfile_idx);
             }
 
-            /* Account for subfiles with uniform segments */
-            if (!is_first && !is_last) {
-                bool thin_uniform_section = false;
+            if (last_subfile_idx < first_subfile_idx) {
+                /*
+                 * This can also happen when the subfile with the final
+                 * data segment has a smaller subfile index than the
+                 * subfile with the first data segment and the current
+                 * subfile index falls between the two.
+                 */
+                thin_uniform_section = thin_uniform_section || ((last_subfile_idx < subfile_idx) &&
```

Comment:
Related to the above comment - I don't think you need the || here

Reply:
Agreed. This was likely a leftover from fixing a bug a long time ago.